src/nlmsg.rs

```diff
@@ -120,6 +148,26 @@ impl NlmsgMut {
         Self::new(buf)
     }
 
+    pub fn nested(ty: u16) -> Self {
+        // 12 => sizeof(ty) + sizeof(nested_attr_ty) + sizeof(nested_attr_data_u32)
+        let mut buffer = Self(BytesMut::with_capacity(12));
+
+        // len of NlAttr will be updated by finish_nested
+        buffer.0.put_slice(bytemuck::bytes_of(&NlAttr {
+            len: 0,
+            ty: ty | libc::NLA_F_NESTED as u16,
+        }));
+        buffer
+    }
+
+    pub fn finish_nested(&mut self, parent: &mut NlmsgMut) {
+        let len = self.0.len();
+        let header: &mut NlAttr =
+            bytemuck::from_bytes_mut(&mut self.0[..core::mem::size_of::<NlAttr>()]);
+        header.len = len as u16;
+        parent.0.put_slice(self.0.as_ref());
```

Comment:
The design here means extra allocation and copy. Perhaps experiment with `BytesMut::split/unsplit` to see if you can just reuse the buffer?

Reply:
I can do that, but the only downside I see is that there's no guarantee the nested buffer will be "finished". It can therefore result in invalid messages being sent to the kernel.

In the current state of the code, if someone forgets to call `finish_nested`, it won't break the message because no byte is added to the "parent" Nlmsg. I could also add a boolean state to the `NlmsgMut` structure, something like `nested_message_in_progress`, and if the user calls `finish` without calling `finish_nested`, an error is returned. It would be a breaking change for the API.

I guess there's a line to be drawn between performance and ensuring the validity of netlink messages. You tell me.

Reply:
If you split_off a `BytesMut`, then the two `BytesMut` will not interfere with each other. They are independent, just like two separately created `BytesMut`. The `unsplit` will try to just adjust the pointer without copy, but it can bail to just extend if the two `BytesMut` are no longer continuous in memory (e.g. if you push something to the first part, or if you push beyond capacity in the second part and causes reallocation).

Reply:
Alright I didn't know that, I will try to experiment with it then